HDPath: add ancestorByDepth(), ancestorByIndex()

msgilligan opened the pull request.

schildbach commented on core/src/main/java/org/bitcoinj/crypto/HDPath.java:
```diff
@@ -444,13 +456,42 @@ public List<HDPath> ancestors() {
                 .collect(StreamUtils.toUnmodifiableList());
     }
 
+    public HDPath ancestorByIndex(int index) {
```
Can you add JavaDoc? I'd like know what this method is supposed to do.

msgilligan replied:
Done!

schildbach commented on core/src/main/java/org/bitcoinj/crypto/HDPath.java:
```diff
@@ -444,13 +456,42 @@ public List<HDPath> ancestors() {
+    }
+
+    // Depth of 0, returns empty path, Depth of d returns ancestorByIndex(d-1)
+    public HDPath ancestorByDepth(int depth) {
```
Can you add JavaDoc? I'd like know what this method is supposed to do.

msgilligan replied:
Done!

schildbach commented on core/src/test/java/org/bitcoinj/crypto/HDPathTest.java:
```diff
@@ -77,6 +77,23 @@ public void testParent() {
+        HDPath path1 = HDPath.parsePath("m/0H/1H");
+
+        assertEquals(HDPath.parsePath("m/"), path1.ancestorByDepth(0));
+        assertEquals(HDPath.parsePath("m/0H"),  path1.ancestorByDepth(1));
+        assertEquals(HDPath.parsePath("m/0H/1H"),  path1.ancestorByDepth(2));
```
One space too many in these two lines.

msgilligan replied:
Fixed.

schildbach commented on core/src/main/java/org/bitcoinj/crypto/HDPath.java:
```diff
@@ -444,13 +456,42 @@ public List<HDPath> ancestors() {
+                : this.ancestorByIndex(depth - 1);
+    }
+
+    private List<ChildNumber> subListInternal(int to) {
+        return childNumbers.subList(0, to);
+    }
```
This refactoring (including the usages) could perhaps be in its own commit, to clear this one up a bit.

msgilligan replied:
Fixed.

schildbach reviewed (changes requested):
I suspect this PR could be split into 2 commits, and also the commit description could briefly explain what the two new methods do.

msgilligan commented:
> I suspect this PR could be split into 2 commits, and also the commit description could briefly explain what the two new methods do.

I split it into 4 commits and force-pushed. I added JavaDoc for the new methods, but just now noticed that you asked for descriptions of the new methods in the commit descriptions. Let me know if you still want that in addition to the JavaDoc.

I also added tests for index out-of-bounds conditions.

schildbach commented:
Merged.
